### PR TITLE
feat(Surveys): Backend changes to associate actions with a Survey

### PR DESCRIPTION
**Pull request by Phanatic**
## Problem

This PR enables the posthog web application to associate actions with a given Survey.

## Changes

The Survey Create/Patch APIs now accept a payload of the format, 
``` json
{
    "conditions": {
        "actions": {
            "values": [
                {
                    "name": "user_subscribed action"
                }
            ]
        }
    }
}
```
and upon receiving this payload, we will then try to match the `name` of each of the `values` within the `actions` property and find associated Actions defined on their project. Once this is created, we will then setup a `many-to-many` association between an Action and its Survey.

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

## Does this work well for both Cloud and self-hosted?

Yes

## How did you test this code?

1. Manually locally
2. Automated tests, see PR

### CODE REVIEW

**Comment by jurajmajerik** [comment]
Looking good! One possible issue (might be posthog-js related):
1. Create a survey with an action
2. Trigger the action and the survey pops up
3. Remove the action from the survey
4. The survey is still popping up

image

File: posthog/api/survey.py [hunk 1]
```diff
@@ -469,11 +533,23 @@ class Meta:
...
+        actions = survey.actions.all()
+        if len(actions) > 0:
+            # action names can change between when the survey is created and when its retrieved.
+            # update the actionNames in the response from the real names of the actions as defined
+            # in data management.
+            if survey.conditions is None:
+                survey.conditions = {}
+
+            survey.conditions["actions"] = {"values": ActionSerializer(actions, many=True).data}
+            # survey.conditions["actions"] =
```
Comment (dmarticus) [inline posthog/api/survey.py:546]:
```suggestion
```

File: posthog/api/test/test_survey.py [hunk 2]
```diff
@@ -926,6 +926,16 @@ def test_inactive_surveys_disables_internal_targeting_flag(self):
...
+    def test_options_unauthenticated(self):
+        unauthenticated_client = Client(enforce_csrf_checks=True)
+        unauthenticated_client.logout()
+        request_headers = {"HTTP_ACCESS_CONTROL_REQUEST_METHOD": "GET", "HTTP_ORIGIN": "*", "USER_AGENT": "Agent 008"}
+        response = unauthenticated_client.options(
+            "/api/surveys", data={}, follow=False, secure=False, headers={}, **request_headers
+        )
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.headers["Access-Control-Allow-Origin"], "*")
+
```
Comment (dmarticus) [inline posthog/api/test/test_survey.py:938]: this feels not related?  It's a good test, but why add it?

File: posthog/api/test/test_survey.py [hunk 3]
```diff
@@ -2036,15 +2233,81 @@ def _get_surveys(
...
+                            }
+                        },
+                        "appearance": None,
+                        "start_date": None,
+                        "end_date": None,
+                        "current_iteration": None,
+                        "current_iteration_start_date": None,
+                    }
+                ],
+            )
```
Comment (dmarticus) [inline posthog/api/test/test_survey.py:2310]: great job on these tests.

**Review by dmarticus — APPROVED** [review]

File: posthog/api/test/test_survey.py (hunk 2 above)
Comment (Phanatic) [reply posthog/api/test/test_survey.py:938]: I had to move this test over here so I can merge `master`, it already existed in this file.

**Pull request merged**